afpacket/src/sockets.rs

```diff
@@ -0,0 +1,181 @@
+#![deny(missing_docs)]
+
+use libc;
+use std::{
+    io,
+    ptr,
+    mem::{self, MaybeUninit},
+    ffi::CStr};
+use crate::linux;
+
+/// Represents a link-local address.
+/// At this time, it's not particularly useful.
+pub struct Addr {
+    inner: libc::sockaddr_storage,
+    len: libc::socklen_t,
+}
+
+/// Represents an unbound `AF_PACKET` socket.  At this phase of a socket's lifecycle, it can be
+/// configured.
+pub struct Socket {
+    fd: libc::c_int,
+}
+
+/// Represents a bound `AF_PACKET` socket. At this phase of a socket's lifecycle, it can be read
+/// to/written from.
+pub struct BoundSocket {
+    fd: libc::c_int,
+    send_addr: libc::sockaddr_ll,
+}
+
+impl Socket {
+    /// Creates a new unbound socket.
+    pub fn new() -> io::Result<Self> {
+        // This block is unsafe because it invokes a syscall. It is safe because it checks the
+        // syscall return value and returns an Err if the sycall fails. It does not interact with
+        // Rust-owned memory, so it's safe as per the borrow checker's rules.
+        let fd = unsafe {
+            // Resources:
+            // https://beej.us/guide/bgnet/html/multi/syscalls.html#socket
+            // man 7 packet
+            let fd = libc::socket(libc::AF_PACKET, libc::SOCK_RAW, libc::ETH_P_ALL.to_be());
+            if fd < 0 {
+                return Err(io::Error::last_os_error())
+            }
+            fd
+        };
+        Ok(Self { fd })
+    }
+
+    /// Binds the socket to a network interface. This function consumes the `Socket` instance, as
+    /// no more configuration options may be safely changed.
+    pub fn bind(self, iface: impl AsRef<CStr>) -> io::Result<BoundSocket> {
+        // This block is unsafe because it invokes syscalls. It is safe because the syscall return
+        // values are checked at each syscall and returned as an Err value. This block should be
+        // tested with miri to confirm that the use of the &CStr it takes as input is correct.
+        let send_addr = unsafe {
+            // get the index of the interface
+            let mut ifr : linux::ifreq = MaybeUninit::zeroed().assume_init();
+            ptr::copy_nonoverlapping(iface.as_ref().as_ptr(),
+                ifr.ifr_ifrn.ifrn_name.as_mut_ptr(), libc::IFNAMSIZ);
+            // ioctl(SIOCGIFINDEX) fills in the index field of the ifreq object
+            // Resources:
+            // man 7 netdevice
+            let err = libc::ioctl(self.fd, linux::SIOCGIFINDEX, &ifr);
+            if err < 0 {
+                return Err(io::Error::last_os_error())
+            }
+
+            // bind the socket
+            let mut ll : libc::sockaddr_ll = MaybeUninit::zeroed().assume_init();
+            ll.sll_family = libc::AF_PACKET as libc::c_ushort;
+            ll.sll_ifindex = ifr.ifr_ifru.ifru_ivalue; // expanded from `ifr_ifindex` in kernel headers
```

Comment:
To paraphrase for my own understanding, we use `ioctl` to determine the _physical_ interface of the thing we're trying to bind to, and now we're setting that on the actual `sockaddr_ll` that we're binding to. I feel like I'm forced to conclude that this some representation of the port number, since there's no other information transferred from the `iface` argument right?

Is there also a reason we aren't setting the other fields on the `sockaddr_ll` struct?

Reply:
I answered my second question. `bind` populates the `sockaddr_ll` and is given to a new `BoundSocket`.

Reply:
Should we resolve this then?

Reply:
If my guess that `ifindex` represents port number, then yes!

Reply:
It's not really a "port number" in the same sense as a TCP/UDP socket. It should be thought of as the NIC's index.

For instance:
```
moosnat@asimov:~$ ip link
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN mode DEFAULT group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
2: enp4s0: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc fq_codel state DOWN mode DEFAULT group default qlen 1000
    link/ether e0:3f:49:2c:3d:d5 brd ff:ff:ff:ff:ff:ff
3: wlp3s0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP mode DORMANT group default qlen 1000
    link/ether 54:27:1e:07:1b:cd brd ff:ff:ff:ff:ff:ff
```

The NIC index is displayed in the output.